File: core/shared/src/test/scala/diplomacy/variant/standard/rule/MovementOrderGeneratorSpec.scala
package diplomacy.variant.standard.rule

import diplomacy.UnitSpec
import diplomacy.board._
import diplomacy.variant.standard.map
import diplomacy.variant.standard.map.Power
import diplomacy.variant.standard.map.Keywords._
import diplomacy.variant.standard.board.{ Turn => T }
import diplomacy.variant.standard.board.Keywords._
import diplomacy.variant.standard.board.Implicits._
import diplomacy.variant.standard.rule.Keywords._
import diplomacy.variant.standard.rule.Implicits._

class MovementOrderGeneratorSpec extends UnitSpec {
  val generator = new MovementOrderGenerator[T, map.Power]
  val board = {
    new Board[T, map.Power](
      map.map,
      1901.Spring - Movement,
      Set(
        DiplomacyUnit(France, Army, Spa),
        DiplomacyUnit(France, Fleet, GoL)
      ), Map(), Map()
    )
  }

  "A movement-order-generator" should {
    "use Hold as a default order." in {
      val $ = StandardRuleOrderHelper(board)
      import $._
      generator.defaultOrder(board)(A(Spa)) should be(A(Spa).hold())
      generator.defaultOrder(board)(F(GoL)) should be(F(GoL).hold())
    }
  }
  it when {
    "there are no units" should {
      "use an empty to skip the movement phase." in {
        val emptyBoard = {
          new Board[T, map.Power](
            map.map, 1901.Spring - Movement, Set(), Map(), Map()
          )
        }
        generator.ordersToSkipPhase(emptyBoard) should be(Some(Set()))
      }
    }

    "there are some units" can {
      "not skip the movement phase." in {
        generator.ordersToSkipPhase(board) should be(None)
      }
    }
  }
}

File: test/standardRule/movement-order-generator-spec.ts
import * as chai from "chai"
import { MovementOrderGenerator } from "./../../src/standardRule/movement-order-generator"
import { standardRule } from "./../../src/standardRule"
import { rule } from "./../../src/rule"
import { standardMap } from "./../../src/standardMap"
import { standardBoard } from "./../../src/standardBoard"

const {
  Board, Unit, MilitaryBranch, State, Result, Phase, Dislodged, ProvinceStatus, StandardRuleHelper: Helper
} = standardRule
declare type Power = standardMap.Power
const { locations: $, Power, map } = standardMap
const { Season, Turn } = standardBoard
const { Army, Fleet } = MilitaryBranch
const { Movement } = Phase

const should = chai.should()

const generator = new MovementOrderGenerator()

describe("A MovementOrderGenerator", () => {
  const board = new Board<Power>(
    map, new State(new Turn(1901, Season.Spring), Movement),
    [
      new Unit(Army, $.Spa, Power.France),
      new Unit(Fleet, $.GoL, Power.France)
    ],
    [], []
  )
  const $$ = new Helper(board)
  it("uses Hold as a default order", () => {
    generator.defaultOrderOf(board, $$.A($.Spa).unit).should.deep.equal($$.A($.Spa).hold())
    generator.defaultOrderOf(board, $$.F($.GoL).unit).should.deep.equal($$.F($.GoL).hold())
  })
  describe("when there are no units", () => {
    const board = new Board(
      map, new State(new Turn(1901, Season.Spring), Movement),
      [], [], []
    )
    it("uses an empty set to skip the movement phase", () => {
      const orders = generator.ordersToSkipPhase(board)
      should.not.equal(orders, null)
      if (orders) {
        Array.from(orders).should.have.deep.members([])
      }
    })
  })
  describe("when there are some units", () => {
    it("does not skip the movement phase", () => {
      should.equal(generator.ordersToSkipPhase(board), null)
    })
  })
})
